### ISSUE TITLE
Assertion `d.nbDims >= 3' failed with int8 for Tensorflow TensorRT integration

### ISSUE THREAD

**Issue by dhingratul** [tensorflow/tensorflow#22514]
Please go to Stack Overflow for help and support:

https://stackoverflow.com/questions/tagged/tensorflow

If you open a GitHub issue, here is our policy:

1. It must be a bug, a feature request, or a significant problem with documentation (for small docs fixes please send a PR instead).
2. The form below must be filled out.
3. It shouldn't be a TensorBoard issue. Those go [here](https://github.com/tensorflow/tensorboard/issues).

**Here's why we have that policy**: TensorFlow developers respond to issues. We want to focus on work that benefits the whole community, e.g., fixing bugs and adding features. Support only helps individuals. GitHub also notifies thousands of people when issues are filed. We want them to see you communicating an interesting problem, rather than being redirected to Stack Overflow.

------------------------

### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:n/a
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: ubuntu 16.04
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: n/a
- **TensorFlow installed from (source or binary)**:source 
- **TensorFlow version (use command below)**: r1.11
- **Python version**: 3.5
- **Bazel version (if compiling from source)**:0.16.1
- **GCC/Compiler version (if compiling from source)**: 5.4.0
- **CUDA/cuDNN version**: 9/7.1
- **GPU model and memory**: 1080ti/11gb
- **Exact command to reproduce**:

```
import tensorflow as tf
from tensorflow.contrib import tensorrt as trt

def load_graph(frozen_graph_filename):
    # We load the protobuf file from the disk and parse it to retrieve the
    # unserialized graph_def
    with tf.gfile.GFile(frozen_graph_filename, "rb") as f:
        graph_def = tf.GraphDef()
        graph_def.ParseFromString(f.read())

    # Then, we can use again a convenient built-in function to import a graph_def into the
    # current default Graph
    with tf.Graph().as_default() as graph:
        tf.import_graph_def(
            graph_def,
            name='', #DEBUG
        )
    return graph

fid = "model.pb"
output_nodenames = 'out1,out2,out3'
output_node = list(output_nodenames.split(","))
g = load_graph(fid)
with tf.Session(graph=g) as sess:
    trt_graph = trt.create_inference_graph(
    input_graph_def=tf.get_default_graph().as_graph_def(),
    outputs=output_node,
    max_batch_size=99999,
    max_workspace_size_bytes=1 << 25,
    precision_mode="INT8",  # TRT Engine precision "FP32","FP16" or "INT8"
    minimum_segment_size=2  # minimum number of nodes in an engine
    )
    with tf.gfile.GFile("trt.pb", "wb") as f:
        f.write(trt_graph.SerializeToString())

g2 = load_graph("trt.pb")
with tf.Session(graph=g2) as sess:
    """Run given calibration graph multiple times."""
    num_samples = 10
    np.random.seed(0)
    ip1_data = np.random.rand(num_samples,700,800,6).astype(np.float32)
    ip1 = g2.get_tensor_by_name("ip1:0")

    ip2_data = np.random.rand(4).astype(np.float32)
    ip2 = g2.get_tensor_by_name("ip2:0")

    ip3_data = np.random.rand(20000,6).astype(np.float32)
    ip3 = g2.get_tensor_by_name("ip3:0")

    ip4_data = np.random.rand(20000,4).astype(np.float32)
    ip4 = g2.get_tensor_by_name("ip4:0")

    out1 = g2.get_tensor_by_name("out1:0")
    out2 = g2.get_tensor_by_name("out2:0")
    out3 = g2.get_tensor_by_name("out3:0")
# run over real calibration data here, we are mimicking a calibration set of
# 30 different batches. Use as much calibration data as you want
    for i in range(num_samples):
        val = sess.run([out1, out2, out3], feed_dict={ip1:ip1_data[i], ip2:ip2_data, ip3:ip3_data, ip4:ip4_data})
```

You can collect some of this information using our environment capture script:

https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh

You can obtain the TensorFlow version with

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

### Describe the problem
Describe the problem clearly here. Be sure to convey here why it's a bug in TensorFlow or a feature request.

### Source code / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

```
2018-09-25 17:20:14.600365: I tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc:577] Starting calibration thread on device 0, Calibration Resource @ 0x7f7fb4001730
2018-09-25 17:20:14.600539: I tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc:577] Starting calibration thread on device 0, Calibration Resource @ 0x7f7fac071a40
2018-09-25 17:20:14.973555: I tensorflow/contrib/tensorrt/kernels/trt_engine_op.cc:577] Starting calibration thread on device 0, Calibration Resource @ 0x7f7fa4004fe0
python: helpers.cpp:56: nvinfer1::DimsCHW nvinfer1::getCHW(const nvinfer1::Dims&): Assertion `d.nbDims >= 3' failed.
Aborted (core dumped)
```

**Comment by samikama** [comment]
@dhingratul ,

do you have a reproducer? I guess what you put above is missing model.pb. As a side note, unless you have a tiny model, setting max batch size to 99999 is a bad idea.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @samikama: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by samikama** [comment]
@dhingratul one option could be playing with segment size option to leave the problematic engines out, if possible. Your problem seem to be due to data dependent shapes in the graph. TFTRT can not properly handle the data dependent shapes yet. We are working on it.

**Comment by dhingratul** [comment]
@samikama All the shapes in the graph, as far as I can see in tensorboard are fixed. How do i "leave problematic engines out"?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @samikama: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by benjamintanweihao** [comment]
@dhingratul I think @samikama means setting the `minimum_segment_size` to a larger number like `50`. However, that doesn't work for me. :man_shrugging:

**Comment by dhingratul** [comment]
@benjamintanweihao I have varied it from 1 to 9999, doesn't work.

**Comment by benjamintanweihao** [comment]
@dhingratul Yeah, I have created an issue https://github.com/tensorflow/tensorflow/issues/24032 for something similar to what you are facing too.

**Comment by blackarrow3542** [comment]
I'm also having this error with INT8 Keras resnet50. 

Code to reproduce the error:
https://github.com/jeng1220/KerasToTensorRT/blob/master/tftrt_resnet_example.py
Change line 66:
`    y = np.empty((num_tests, self.y_tensor.shape[1]), np.float32)`
to     `y = np.empty((num_tests, 1000), np.float32)`
line 108   `tftrt_engine = TftrtEngine(frozen_graph, batch_size, 'FP32')`
to  `tftrt_engine = TftrtEngine(frozen_graph, batch_size, 'INT8')`
I tried tensorflow-gpu1.12/tensorflow-gpu1.11 and TensorRT 4.0
GTX 1080ti
CUDA 9.0 cudnn 7.0
Driver Version: 390.87 
FP32 works for the keras resnet50.

**Comment by blackarrow3542** [comment]
Seems like there is a related issue at:
https://github.com/tensorflow/models/issues/5061
with  the int8 tensorrt example for resnet50 in tensorflow
https://github.com/tensorflow/models/blob/master/research/tensorrt/tensorrt.py
not working. 

Another code for resnet50 with INT8 works:  
https://github.com/tensorflow/models/issues/5061#issuecomment-419235783
> That's similar to this one and INT8 works perfectly. Here is the link:
> https://developer.download.nvidia.com/devblogs/tftrt_sample.tar.xz
> I found it in this link:
> https://devblogs.nvidia.com/tensorrt-integration-speeds-tensorflow-inference/
> Hope it can solve the problem for you as well.
> Please let me know if you have any question about it.

**Comment by samikama** [comment]
Is this still there with TRT5.0?

**Comment by samikama** [comment]
@trevor-m @pooyadavoodi  Can you PTAL

**Comment by pooyadavoodi** [comment]
I believe this is fixed in a more recent version of TF.

I just tried the script that @blackarrow3542  linked in https://github.com/tensorflow/tensorflow/issues/22514#issuecomment-443076374
I tested with TF 1.14, had to make some changes to adjust to the new TF-TRT API.
The test passed.

This is the log:

```
2019-06-19 23:58:57.567682: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1326] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 22664 MB memory) -> physical GPU (device: 0, name: TITAN RTX, pci bus id: 0000:02:00.0, compute capability: 7.5)
2019-06-19 23:58:57.966968: I tensorflow/compiler/tf2tensorrt/segment/segment.cc:460] There are 4 ops of 3 different types in the graph that are not converted to TensorRT: Softmax, Placeholder, NoOp, (For more information see https://docs.nvidia.com/deeplearning/dgx/tf-trt-user-guide/index.html#supported-ops).
2019-06-19 23:58:58.008931: I tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc:733] Number of TensorRT candidate segments: 1
2019-06-19 23:58:58.024111: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudart.so.10.1
2019-06-19 23:58:58.109276: I tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc:835] TensorRT node TRTEngineOp_0 added for segment 0 consisting of 503 nodes succeeded.
2019-06-19 23:58:58.129599: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:739] Optimization results for grappler item: tf_graph
2019-06-19 23:58:58.129650: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:741]   constant folding: Graph size after: 555 nodes (0), 570 edges (0), time = 44.491ms.
2019-06-19 23:58:58.129661: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:741]   layout: Graph size after: 560 nodes (5), 572 edges (2), time = 27.786ms.
2019-06-19 23:58:58.129669: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:741]   constant folding: Graph size after: 557 nodes (-3), 572 edges (0), time = 43.335ms.
2019-06-19 23:58:58.129676: I tensorflow/core/grappler/optimizers/meta_optimizer.cc:741]   TensorRTOptimizer: Graph size after: 55 nodes (-502), 54 edges (-518), time = 274.138ms.
2019-06-19 23:58:59.270390: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1640] Found device 0 with properties:
name: TITAN RTX major: 7 minor: 5 memoryClockRate(GHz): 1.77
pciBusID: 0000:02:00.0
2019-06-19 23:58:59.270441: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudart.so.10.1
2019-06-19 23:58:59.270465: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcublas.so.10
2019-06-19 23:58:59.270476: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcufft.so.10
2019-06-19 23:58:59.270487: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcurand.so.10
2019-06-19 23:58:59.270499: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcusolver.so.10
2019-06-19 23:58:59.270509: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcusparse.so.10
2019-06-19 23:58:59.270520: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudnn.so.7
2019-06-19 23:58:59.271632: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1763] Adding visible gpu devices: 0
2019-06-19 23:58:59.271664: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1181] Device interconnect StreamExecutor with strength 1 edge matrix:
2019-06-19 23:58:59.271671: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1187]      0
2019-06-19 23:58:59.271677: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1200] 0:   N
2019-06-19 23:58:59.273011: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1326] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 22664 MB memory) -> physical GPU (device: 0, name: TITAN RTX, pci bus id: 0000:02:00.0, compute capability: 7.5)
2019-06-19 23:58:59.471499: I tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc:716] Starting calibration thread on device 0, Calibration Resource @ 0x7fe9e4008c40
2019-06-20 00:01:06.687922: I tensorflow/compiler/tf2tensorrt/utils/trt_resources.cc:25] Destroying Calibration Resource
 Calibrator = 0x7fe9e40023c0
 Builder    = 0
 Engine     = 0x7fe38668c5c0
 Logger     = 0x7fe9e4008cc0
 Allocator  = 0x7fe9e4008770
 Thread     = 0x7fe9e4008d00

2019-06-20 00:01:07.445518: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1640] Found device 0 with properties:
name: TITAN RTX major: 7 minor: 5 memoryClockRate(GHz): 1.77
pciBusID: 0000:02:00.0
2019-06-20 00:01:07.445569: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudart.so.10.1
2019-06-20 00:01:07.445589: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcublas.so.10
2019-06-20 00:01:07.445600: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcufft.so.10
2019-06-20 00:01:07.445609: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcurand.so.10
2019-06-20 00:01:07.445620: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcusolver.so.10
2019-06-20 00:01:07.445630: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcusparse.so.10
2019-06-20 00:01:07.445639: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudnn.so.7
2019-06-20 00:01:07.446709: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1763] Adding visible gpu devices: 0
2019-06-20 00:01:07.446740: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1181] Device interconnect StreamExecutor with strength 1 edge matrix:
2019-06-20 00:01:07.446747: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1187]      0
2019-06-20 00:01:07.446751: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1200] 0:   N
2019-06-20 00:01:07.448045: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1326] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 22664 MB memory) -> physical GPU (device: 0, name: TITAN RTX, pci bus id: 0000:02:00.0, compute capability: 7.5)
2019-06-20 00:01:07.637419: I tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc:632] Building a new TensorRT engine for import/TRTEngineOp_0 input shapes: [[1,224,224,3]]
Keras time 2.380934238433838
Tensorflow time 0.5712037086486816
PASSED
TFTRT time 67.95880579948425
PASSED
```

To use the new TF-TRT API, you would have to do something like this:
```
    converter = tftrt.TrtGraphConverter(
      input_graph_def=graph.frozen,
      nodes_blacklist=graph.y_name,
      max_batch_size=batch_size,
      max_workspace_size_bytes=1 << 30,
      precision_mode=precision,
      minimum_segment_size=2)
    tftrt_graph = converter.convert()
    batch_size = 1
    img_shape = (224, 224, 3)
    x = np.random.random_sample((batch_size,
      img_shape[0], img_shape[1], img_shape[2]))
    tftrt_graph = converter.calibrate(
      fetch_names=graph.y_name,
      num_runs=10,
      feed_dict_fn=lambda: {graph.x_name[0] + ':0': x[0: batch_size]})
```

**Comment by aaroey** [comment]
This is fixed in 1.14.
